test/performance/odp_ipsecfwd.c

```diff
@@ -0,0 +1,1307 @@
+		pktio_param.out_mode = config->mode == 0U ? ODP_PKTOUT_MODE_QUEUE :
+							    ODP_PKTOUT_MODE_DIRECT;
+		pktio->handle = odp_pktio_open(pktio->name, config->pktio_pool, &pktio_param);
+
+		if (pktio->handle == ODP_PKTIO_INVALID) {
+			ODPH_ERR("Error opening packet I/O (%s)\n", pktio->name);
+			return false;
+		}
+
+		odp_pktin_queue_param_init(&pktin_param);
+
+		if (config->num_thrs > 1) {
+			pktin_param.hash_enable = true;
+			pktin_param.hash_proto.proto.ipv4_udp = 1U;
+			pktin_param.num_queues = config->num_thrs;
+		}
+
+		if (odp_pktin_queue_config(pktio->handle, &pktin_param) < 0) {
+			ODPH_ERR("Error configuring packet I/O input queues (%s)\n", pktio->name);
+			return false;
+		}
+
+		if (odp_pktio_capability(pktio->handle, &capa) < 0) {
+			ODPH_ERR("Error querying packet I/O capabilities (%s)\n", pktio->name);
+			return false;
+		}
+
+		odp_pktout_queue_param_init(&pktout_param);
+
+		if (config->mode == 0U) {
+			pktio->send_fn = enqueue;
+			pktio->num_tx_qs = 1U;
+		} else {
+			num_tx_qs = MIN((uint32_t)config->num_thrs, MIN(MAX_QUEUES,
+									capa.max_output_queues));
+			pktout_param.num_queues = num_tx_qs;
+			pktout_param.op_mode = ODP_PKTIO_OP_MT_UNSAFE;
+			pktio->send_fn = send;
+			pktio->num_tx_qs = num_tx_qs;
+		}
+
+		if (odp_pktout_queue_config(pktio->handle, &pktout_param) < 0) {
+			ODPH_ERR("Error configuring packet I/O output queues (%s)\n", pktio->name);
+			return false;
+		}
+
+		if (config->mode == 0U) {
+			if (odp_pktout_event_queue(pktio->handle, &pktio->out_ev_q, 1) != 1) {
+				ODPH_ERR("Error querying packet I/O output event queue (%s)\n",
+					 pktio->name);
+				return false;
+			}
+		} else {
+			if (odp_pktout_queue(pktio->handle, pktio->out_dir_qs, pktio->num_tx_qs)
+			    != (int)pktio->num_tx_qs) {
+				ODPH_ERR("Error querying packet I/O output queue (%s)\n",
+					 pktio->name);
+				return false;
+			}
+		}
+
+		odp_pktio_config_init(&pktio_config);
+
+		if (odp_pktio_config(pktio->handle, &pktio_config) < 0) {
+			ODPH_ERR("Error configuring packet I/O extra options (%s)\n", pktio->name);
+			return false;
+		}
+
+		(void)odp_pktio_mac_addr(pktio->handle, &pktio->src_mac, sizeof(pktio->src_mac));
+
+		if (odp_pktio_start(pktio->handle) < 0) {
+			ODPH_ERR("Error starting packet I/O (%s)\n", pktio->name);
+			return false;
+		}
+	}
+
+	return true;
+}
+
+static odp_bool_t setup_fwd_table(prog_config_t *config)
+{
+	fwd_entry_t *fwd_e;
+
+	config->fwd_tbl = odph_iplookup_table_create(PROG_NAME, 0U, 0U, sizeof(fwd_entry_t *));
+
+	if (config->fwd_tbl == NULL) {
+		ODPH_ERR("Error creating forwarding table\n");
+		return false;
+	}
+
+	for (uint32_t i = 0U; i < config->num_fwds; ++i) {
+		fwd_e = &config->fwd_entries[i];
+
+		if (odph_iplookup_table_put_value(config->fwd_tbl, &fwd_e->prefix, &fwd_e) < 0) {
+			ODPH_ERR("Error populating forwarding table\n");
+			return false;
+		}
+	}
+
+	return true;
+}
+
+static inline odp_ipsec_sa_t *get_in_sa(odp_packet_t pkt)
```

Comment:
I though it was slightly odd that these return a pointer to the handle instead of the handle itself, but maybe it does not really matter.

Reply:
Since this does not really matter, at least now, we can leave it as it is.